tests/test_charm.py

```diff
@@ -322,6 +372,33 @@ def test_k8s_scrape_config_can_be_set(self):
         k8s_pod_scrape_config = scrape_config(pod_spec, 'kubernetes-pods')
         self.assertIsNotNone(k8s_pod_scrape_config, 'No k8s pods scrape config found')
 
+    def test_image_tags_with_patch_versions(self):
+        self.harness.set_leader(True)
+        k8s_config = MINIMAL_CONFIG.copy()
+        self.harness.update_config(k8s_config)
+        api_tags = [{'layer': '', 'name': 'latest'}, {'layer': '', 'name': 'v2.2.0-rc.0'}, {'layer': '', 'name': 'v2.2.0-rc.1'}, {'layer': '', 'name': 'v2.2.1'}, {'layer': '', 'name': 'v2.20'}, {'layer': '', 'name': 'v2.20.0'}, {'layer': '', 'name': 'v2.20.0-rc.0'}, {'layer': '', 'name': 'v2.20.0-rc.1'}, {'layer': '', 'name': 'v2.20.1'}, {'layer': '', 'name': 'v2.21.0'}, {'layer': '', 'name': 'v2.21.0-rc.0'}, {'layer': '', 'name': 'v2.21.0-rc.1'}, {'layer': '', 'name': 'v2.3.0'}, {'layer': '', 'name': 'v2.3.1'}, {'layer': '', 'name': 'v2.3.2'}, {'layer': '', 'name': 'v2.4.0'}] # noqa
+        actual_tag = self.harness.charm._get_image_tag(api_tags=api_tags)
+        expected_tag = 'v2.20.1'
+        self.assertEqual(actual_tag, expected_tag)
+
+    def test_image_tags_without_patch_versions(self):
+        self.harness.set_leader(True)
+        k8s_config = MINIMAL_CONFIG.copy()
+        self.harness.update_config(k8s_config)
+        api_tags = [{'layer': '', 'name': 'v2.20'}, {'layer': '', 'name': 'v2.21.0'}, {'layer': '', 'name': 'v2.21.0-rc.0'}, {'layer': '', 'name': 'v2.21.0-rc.1'}, {'layer': '', 'name': 'v2.3.0'}, {'layer': '', 'name': 'v2.3.1'}, {'layer': '', 'name': 'v2.3.2'}, {'layer': '', 'name': 'v2.4.0'}] # noqa
+        actual_tag = self.harness.charm._get_image_tag(api_tags=api_tags)
+        expected_tag = 'v2.20'
+        self.assertEqual(actual_tag, expected_tag)
+
+    def test_no_valid_tags(self):
+        self.harness.set_leader(True)
+        k8s_config = MINIMAL_CONFIG.copy()
+        self.harness.update_config(k8s_config)
+        api_tags = [{'layer': '', 'name': 'latest'}, {'layer': '', 'name': 'v2.2.0-rc.0'}, {'layer': '', 'name': 'v2.2.0-rc.1'}, {'layer': '', 'name': 'v2.2.1'}, {'layer': '', 'name': 'v2.20.0-rc.0'}, {'layer': '', 'name': 'v2.20.0-rc.1'}, {'layer': '', 'name': 'v2.21.0'}, {'layer': '', 'name': 'v2.21.0-rc.0'}, {'layer': '', 'name': 'v2.21.0-rc.1'}, {'layer': '', 'name': 'v2.3.0'}, {'layer': '', 'name': 'v2.3.1'}, {'layer': '', 'name': 'v2.3.2'}, {'layer': '', 'name': 'v2.4.0'}] # noqa
```

Comment:
These feel like they would be easier to read if you did them as:

```
api_tags = [
  {'layer': '', 'name': 'latest'}, 
  {'layer': '', 'name': 'v2.2.0-rc.0'},
  {'layer': '', 'name': 'v2.2.0-rc.1'},
  {'layer': '', 'name': 'v2.2.1'},
  {'layer': '', 'name': 'v2.20.0-rc.0'},
  {'layer': '', 'name': 'v2.20.0-rc.1'},
  {'layer': '', 'name': 'v2.21.0'},
...]
```

And then you can drop the #noqa as well.

Reply:
Good call. I'll make the change.